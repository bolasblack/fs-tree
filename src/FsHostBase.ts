import { Host as IHost, StatsModifyOptions } from './interfaces'
import {
  mkdirp,
  FsImplementation as MkdirpFsImplementation,
} from './mkdirp-then'
import { groupBy, zipObj, mergeRight, compose } from 'ramda'
import { promisify } from 'util'
import sysPath from 'path'
import fs from 'fs'
import {
  FileAlreadyExistException,
  FileDoesNotExistException,
  PathIsDirectoryException,
} from './exceptions'

type NativeFS = typeof fs

export interface FSImpl extends MkdirpFsImplementation {
  readFile: NativeFS['readFile']
  readdir: NativeFS['readdir']
  stat: NativeFS['stat']
  writeFile: NativeFS['writeFile']
  chmod: NativeFS['chmod']
  rename: NativeFS['rename']
  unlink: NativeFS['unlink']
}

export class FsHostBase implements IHost {
  constructor(private _fs: FSImpl) {}

  read(path: string) {
    return promisify(this._fs.readFile.bind(this._fs))(path)
  }

  async readDir(path: string) {
    const entries = await promisify(this._fs.readdir.bind(this._fs))(path)
    const paths = entries.map(entry => sysPath.resolve(path, entry))

    const types = await Promise.all(
      paths.map(async entryPath => {
        return (await this.readStat(entryPath)).isDirectory() ? 'dirs' : 'files'
      }),
    )
    const typeMap = zipObj(paths, types)

    return compose(
      mergeRight({ files: [], dirs: [] }),
      groupBy((path: string) => typeMap[path]),
    )(paths) as {
      files: string[]
      dirs: string[]
    }
  }

  readStat(path: string) {
    return promisify(this._fs.stat.bind(this._fs))(path)
  }

  async overwrite(
    path: string,
    content?: Buffer,
    options?: IHost.create.Options,
  ) {
    const originStat = await this._safeReadStat(path)

    if (!originStat) {
      throw new FileDoesNotExistException(path)
    }

    if (originStat.isDirectory()) {
      throw new PathIsDirectoryException(path)
    }

    await this._write(path, content, options)
  }

  async create(path: string, content: Buffer, options?: IHost.create.Options) {
    if (await this._safeReadStat(path)) {
      throw new FileAlreadyExistException(path)
    }

    await this._write(path, content, options)
  }

  async move(from: string, to: string) {
    const [sourcePathStat, targetPathStat] = await Promise.all([
      this._safeReadStat(from),
      this._safeReadStat(to),
    ])

    if (!sourcePathStat) {
      throw new FileDoesNotExistException(from)
    }

    if (from === to) return

    if (targetPathStat) {
      throw new FileAlreadyExistException(to)
    }

    await this._mkdirToParent(to)
    await promisify(this._fs.rename.bind(this._fs))(from, to)
  }

  async delete(path: string) {
    const stat = await this._safeReadStat(path)

    if (!stat) {
      throw new FileDoesNotExistException(path)
    }

    if (stat.isDirectory()) {
      throw new PathIsDirectoryException(path)
    }

    await promisify(this._fs.unlink.bind(this._fs))(path)
  }

  private async _mkdirToParent(path: string) {
    const parentPath = sysPath.resolve(path, '../')

    try {
      await mkdirp(parentPath, { fs: this._fs })
    } catch (err) {
      if (err.code !== 'EEXIST') throw err
    }
  }

  private async _write(
    path: string,
    content?: Buffer,
    options?: StatsModifyOptions,
  ) {
    await this._mkdirToParent(path)

    if (content != null) {
      await promisify(this._fs.writeFile.bind(this._fs))(path, content, options)
    }

    // memfs.writeFile not support assign file mode
    if (options && options.mode != null) {
      await promisify(this._fs.chmod.bind(this._fs))(path, options.mode)
    }
  }

  private async _safeReadStat(path: string) {
    try {
      return await this.readStat(path)
    } catch (err) {
      if (err.code === 'ENOENT') return null
      throw err
    }
  }
}
