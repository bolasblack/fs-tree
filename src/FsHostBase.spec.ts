import { Volume } from 'memfs'
import { FsHostBase } from './FsHostBase'

describe('FsHostBase', () => {
  describe('#read', () => {
    it('read file content', async () => {
      const [, host] = createHost({ '/a/b': 'content' })
      const resp = await host.read('/a/b')
      expect(resp).toEqual(Buffer.from('content'))
    })
  })

  describe('#readStat', () => {
    it('read stat', async () => {
      const [fs, host] = createHost({ '/a/b': 'content' })
      fs.chmodSync('/a/b', 0o400)
      const resp = await host.readStat('/a/b')
      expect(resp).toEqual(
        expect.customChecker(stat => (stat.mode & 0o777) === 0o400),
      )
    })
  })

  describe('#readDir', () => {
    it('read and organize folder structure', async () => {
      const [, host] = createHost({
        '/a/b': 'content',
        '/c': 'content',
        '/a/d/e': 'content',
      })

      await expect(host.readDir('/')).resolves.toEqual({
        files: ['/c'],
        dirs: ['/a'],
      })

      await expect(host.readDir('/a')).resolves.toEqual({
        files: ['/a/b'],
        dirs: ['/a/d'],
      })
    })
  })

  describe('#overwrite', () => {
    it('throw error if target not exists', async () => {
      const [, host] = createHost({})
      await expect(
        host.overwrite('/file/not/exists', Buffer.from('new content')),
      ).rejects.toMatchSnapshot()
    })

    it('throw error if target is directory', async () => {
      const dirPath = '/some/path/to'
      const [, host] = createHost({ [dirPath + '/file']: 'content' })
      await expect(
        host.overwrite(dirPath, Buffer.from('new content')),
      ).rejects.toMatchSnapshot()
    })

    it('overwrite file content', async () => {
      const filePath = '/some/path/to/file'
      const [fs, host] = createHost({ [filePath]: 'content' })
      await host.overwrite(filePath, Buffer.from('content1'))
      expect(fs.readFileSync(filePath).toString()).toBe('content1')
    })

    it('overwrite file stat', async () => {
      const filePath = '/some/path/to/file'
      const [fs, host] = createHost({ [filePath]: 'content' })

      const fileModeBefore = fs.statSync(filePath).mode
      expect(fileModeBefore & 0o777).toBe(0o666)

      await host.overwrite(filePath, undefined, {
        mode: 0o733,
      })
      const fileModeAfter = fs.statSync(filePath).mode
      expect(fileModeAfter & 0o777).toBe(0o733)
    })

    it('overwrite file with empty content', async () => {
      const filePath = '/some/path/to/file'
      const [fs, host] = createHost({ [filePath]: 'content' })
      await host.overwrite(filePath, Buffer.from(''))
      expect(fs.readFileSync(filePath).toString()).toBe('')
    })
  })

  describe('#create', () => {
    it('throw error if target already exists', async () => {
      const filePath = '/some/path/to/file'
      const [, host] = createHost({ [filePath]: 'content' })
      await expect(
        host.create(filePath, Buffer.from('content1')),
      ).rejects.toMatchSnapshot()
    })

    it('throw error if target is directory', async () => {
      const dirPath = '/some/path/to'
      const [, host] = createHost({ [dirPath + '/file']: 'content' })
      await expect(
        host.create(dirPath, Buffer.from('content')),
      ).rejects.toMatchSnapshot()
    })

    it('create file', async () => {
      const filePath = '/some/path/to/file'
      const [fs, host] = createHost({ '/some/path/other/file': 'content1' })
      await host.create(filePath, Buffer.from('content'))
      expect(fs.readFileSync(filePath).toString()).toBe('content')
    })

    it('create file with stat', async () => {
      const filePath = '/some/path/to/file'
      const [fs, host] = createHost({ '/some/path/other/file': 'content1' })
      await host.create(filePath, Buffer.from('content'), {
        mode: 0o733,
      })
      const fileMode = fs.statSync(filePath).mode
      expect(fileMode & 0o777).toBe(0o733)
    })

    it('create file with empty content', async () => {
      const filePath = '/some/path/to/file'
      const [fs, host] = createHost({ '/some/path/other/file': 'content1' })
      await host.create(filePath, Buffer.from(''))
      expect(fs.readFileSync(filePath).toString()).toBe('')
    })
  })

  describe('#delete', () => {
    it('throw error if target is directory', async () => {
      const dirPath = '/some/path/to'
      const [, host] = createHost({ [dirPath + '/file']: 'content' })
      await expect(host.delete(dirPath)).rejects.toMatchSnapshot()
    })

    it('throw error if target not exists', async () => {
      const [, host] = createHost({})
      await expect(host.delete('/not/existed/path')).rejects.toMatchSnapshot()
    })

    it('delete file', async () => {
      const filePath = '/some/path/to/file'
      const [fs, host] = createHost({ [filePath]: 'content' })
      await host.delete(filePath)
      expect(fs.existsSync(filePath)).toBe(false)
    })
  })

  describe('#move', () => {
    it('throw error if source not exists', async () => {
      const [, host] = createHost({})
      await expect(
        host.move('/not/existed/path', '/not/existed/path1'),
      ).rejects.toMatchSnapshot()
    })

    it('throw error if target is existed', async () => {
      const filePath = '/some/path/to/file'
      const [, host] = createHost({
        [`${filePath}1`]: 'content',
        [`${filePath}2`]: 'content',
      })
      await expect(
        host.move(`${filePath}1`, `${filePath}2`),
      ).rejects.toMatchSnapshot()
    })

    it('move file from source to target', async () => {
      const filePath = '/some/path/to/file'
      const [fs, host] = createHost({ [filePath]: 'content' })
      await host.move(filePath, `${filePath}1`)
      expect(fs.existsSync(filePath)).toBe(false)
      expect(fs.existsSync(filePath + '1')).toBe(true)
    })
  })
})

function createHost(fileStruct: { [path: string]: any } = {}) {
  const fs = Volume.fromJSON(fileStruct)
  const host = new FsHostBase(fs as any)
  return [fs, host] as [typeof fs, FsHostBase]
}
