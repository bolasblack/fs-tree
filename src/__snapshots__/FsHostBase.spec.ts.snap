// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FsHostBase #create throw error if target already exists 1`] = `
FileAlreadyExistException(
  "Path \\"/some/path/to/file\\" already exist."
  { path: '/some/path/to/file' }
)
`;

exports[`FsHostBase #create throw error if target is directory 1`] = `
FileAlreadyExistException(
  "Path \\"/some/path/to\\" already exist."
  { path: '/some/path/to' }
)
`;

exports[`FsHostBase #delete throw error if target is directory 1`] = `
PathIsDirectoryException(
  "Path \\"/some/path/to\\" is a directory."
  { path: '/some/path/to' }
)
`;

exports[`FsHostBase #delete throw error if target not exists 1`] = `
FileDoesNotExistException(
  "Path \\"/not/existed/path\\" does not exist."
  { path: '/not/existed/path' }
)
`;

exports[`FsHostBase #move throw error if source not exists 1`] = `
FileDoesNotExistException(
  "Path \\"/not/existed/path\\" does not exist."
  { path: '/not/existed/path' }
)
`;

exports[`FsHostBase #move throw error if target is existed 1`] = `
FileAlreadyExistException(
  "Path \\"/some/path/to/file2\\" already exist."
  { path: '/some/path/to/file2' }
)
`;

exports[`FsHostBase #overwrite throw error if target is directory 1`] = `
PathIsDirectoryException(
  "Path \\"/some/path/to\\" is a directory."
  { path: '/some/path/to' }
)
`;

exports[`FsHostBase #overwrite throw error if target not exists 1`] = `
FileDoesNotExistException(
  "Path \\"/file/not/exists\\" does not exist."
  { path: '/file/not/exists' }
)
`;
